### PR TITLE
[BE] Username and password register changes

### PR DESCRIPTION
**Pull request by GitVivekHub**
## 📋 Description

JIRA ID: 

Please provide a summary of the change and the motivation behind it. Include relevant context and details.

---

## ✅ Type of Change

- [ ] 🐞 **Bug fix** (non-breaking change which resolves an issue)
- [ ] ✨ **New feature** (non-breaking change which adds functionality)
- [ ] 🔥 **Breaking change** (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🛠 **Refactor** (change that is neither a fix nor a new feature)
- [ ] ⚙️ **Config change** (configuration file or build script updates)
- [ ] 📚 **Documentation** (updates to docs or readme)
- [ ] 🧪 **Tests** (adding new or updating existing tests)
- [ ] 🎨 **UI/UX** (changes that affect the user interface)
- [ ] 🚀 **Performance** (improves performance)
- [ ] 🧹 **Chore** (miscellaneous changes that don't modify src or test files)

---

## ℹ️ Additional Information

Please describe how the changes were tested, and include any relevant screenshots, logs, or other information that provides additional context.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new user registration method allowing users to register with a username and password.
	- Enhanced username generation logic for improved user identification.
	- Password field in registration is now optional, with new validation rules for enhanced security.
	- Added a new endpoint for user registration with username and password in the authentication controller.

- **Bug Fixes**
	- Refined error handling to ensure proper deletion of Keycloak users on registration errors.

### CODE REVIEW

File: src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -119,13 +119,18 @@ export class AuthService {
...
+        body?.phoneNumber?.slice(-4),
       credentials: [
-        // {
-        //   type: 'password',
-        //   value: body?.password,
-        //   temporary: false,
-        // },
+        {
+          type: 'password',
+          value: 'Password@123',
```
Comment (manojLondhe) [inline src/modules/auth/auth.service.ts:131]: Use env config for setting default password value

REGISTRATION_DEFAULT_PWD

Comment (manojLondhe) [reply src/modules/auth/auth.service.ts:131]: Accept pwd from input, if input not found set default pwd

**Review by manojLondhe — CHANGES_REQUESTED** [review]

File: src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -119,13 +119,18 @@ export class AuthService {
       enabled: 'true',
       firstName: body?.firstName,
       lastName: body?.lastName,
-      username: body?.phoneNumber,
+      username:
+        body?.firstName +
```
Comment (ManojNathIC) [inline src/modules/auth/auth.service.ts:123]: @GitVivekHub could you please use a function to generate the username and consider the extra spaces in data, you can use trim or replace it with an underscore 
**Ref** below function  and as discussed yesterday make replace aadhaar with the phone number in this function too
https://github.com/PSMRI/beneficiary-backend/blob/main/src/modules/users/users.service.ts#L704

File: src/modules/auth/dto/register.dto.ts [hunk 3]
```diff
@@ -20,12 +26,7 @@ export class RegisterDTO {
...
-  // @ApiProperty({})
-  // @IsNotEmpty({ message: 'Password is required' })
-  // @MinLength(8, { message: 'Password must be at least 8 characters long' })
-  // @Matches(/^(?=.*[A-Z])(?=.*[a-z])(?=.*\d)(?=.*[\W_]).*$/, {
-  //   message:
-  //     'Password must contain at least one uppercase letter, one lowercase letter, one number, and one special character',
-  // })
-  // password: string;
+  @ApiProperty({})
+  @IsOptional()
```
Comment (ManojNathIC) [inline src/modules/auth/dto/register.dto.ts:30]: Please add pwd validation like suggested by coderabbit

**Review by ManojNathIC — CHANGES_REQUESTED** [review]

File: src/modules/auth/auth.service.ts [hunk 4]
```diff
@@ -98,6 +98,59 @@ export class AuthService {
     }
   }
 
+  public async registerWithUsernamePassword(body) {
+    try {
+      let wallet_api_url = process.env.WALLET_API_URL;
+      // Step 1: Check if mobile number exists in the database
+      await this.checkMobileExistence(body?.phoneNumber);
```
Comment (manojLondhe) [inline src/modules/auth/auth.service.ts:105]: we may have multiple users entering same number?

File: src/modules/auth/auth.service.ts [hunk 5]
```diff
@@ -136,6 +187,39 @@ export class AuthService {
     };
   }
 
+  private prepareUserDataV2(body) {
+    const trimmedFirstName = body?.firstName?.trim();
+    const trimmedLastName = body?.lastName?.trim();
+    const trimmedPhoneNumber = body?.phoneNumber?.trim();
+    const password = body?.password || process.env.SIGNUP_DEFAULT_PASSWORD;
```
Comment (ManojNathIC) [inline src/modules/auth/auth.service.ts:194]: @GitVivekHub Please use spaces trim forn password

**Review by ManojNathIC — CHANGES_REQUESTED** [review]

**Review by ManojNathIC — APPROVED** [review]

**Pull request merged**